### PR TITLE
wifi: Add full AP mode support (using hostapd)

### PR DESCRIPTION
**Pull request by nxf58150**
Added support of setting up AP with hostapd. 
Once flag CONFIG_WIFI_NM_HOSTAPD_AP is enabled, softAP related commands in l2 wifi_shell.c will be handled by hostapd.
Increase max value of CONFIG_WIFI_SHELL_MAX_AP_STA from 5 to 8. The softAP of NXP wifi chip can support up to 8 stations.

### CODE REVIEW

File: modules/hostap/src/supp_main.c [hunk 1]
```diff
@@ -523,6 +585,456 @@ static int register_supplicant_event_socket(struct supplicant_context *ctx)
...
+	bss                 = conf->last_bss;
+	bss->start_disabled = 1;
+	bss->max_num_sta    = 8;
+	bss->dtim_period    = 1;
+	os_strlcpy(conf->bss[0]->iface, ifname, sizeof(conf->bss[0]->iface));
+	bss->logger_stdout_level = HOSTAPD_LEVEL_INFO;
+	bss->logger_stdout       = 0xffff;
+	bss->nas_identifier      = os_strdup("ap.example.com");
+	bss->eap_sim_db          = os_strdup("unix:/tmp/hlr_auc_gw.sock");
+	os_memcpy(conf->country, "US ", 3);
```
Comment (jukkar) [inline modules/hostap/src/supp_main.c:846]: The domain socket is something that is not needed in zephyr.
Also the hardcoded values look wrong.

File: modules/hostap/src/supp_main.c [hunk 2]
```diff
@@ -523,6 +585,456 @@ static int register_supplicant_event_socket(struct supplicant_context *ctx)
...
+		return;
+	}
+
+	eloop_register_timeout(HOSTAPD_CLEANUP_INTERVAL, 0,
+			       hostapd_periodic, interfaces, NULL);
+
+	iface = net_if_get_wifi_uap();
+	/* Allocate and parse configuration for full interface files */
+	dev = net_if_get_device(iface);
+	strncpy(ifname, dev->name, IFNAMSIZ);
```
Comment (jukkar) [inline modules/hostap/src/supp_main.c:993]: We should not use device name as an interface name.

**Comment by jukkar** [comment]
Isn't this depending on https://github.com/zephyrproject-rtos/hostap/pull/15, the west.yml should be updated too in that case.

File: modules/hostap/Kconfig [hunk 3]
```diff
@@ -171,6 +175,9 @@ config WIFI_NM_WPA_SUPPLICANT_WPA3
 config WIFI_NM_WPA_SUPPLICANT_AP
 	bool "AP mode support"
 
+config WIFI_NM_HOSTAPD_AP
+	bool "AP mode hostapd support"
```
Comment (krish2718) [inline modules/hostap/Kconfig:179]: `hostapd based full AP mode` and we can fix `WIFI_NM_WPA_SUPPLICANT_AP` to ` SoftAP mode support based on WPA supplicant`

File: modules/hostap/src/supp_api.c [hunk 4]
```diff
@@ -906,10 +955,151 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
...
+}
+
+int hapd_config_network(struct hostapd_iface *iface,
+			struct wifi_connect_req_params *params)
+{
+	int ret = 0;
+
+	if (!hostapd_cli_cmd_v("set ssid %s", params->ssid)) {
+		goto out;
+	}
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:987]: you can use a macro do the NULL check + goto out, like we did for WPA supplicant.

File: modules/hostap/src/supp_api.c [hunk 5]
```diff
@@ -906,10 +955,151 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
...
+
+	*state = iface->state;
+
+out:
+	k_mutex_unlock(&wpa_supplicant_mutex);
+	return ret;
+}
+
+int hapd_config_network(struct hostapd_iface *iface,
```
Comment (krish2718) [inline modules/hostap/src/supp_api.c:980]: IIUC, the commands itself are common b/w SoftAP and FullAP, so, may be if you write a wrapper to choose wpa_cli_cmd_v vs hostapd_cli_cmd_v, we can reuse the code?

**Review by krish2718 — COMMENTED** [review]
I don't see the dynamic interfaces support for HOSTAPD_AP i.e., monitoring interface state events and accordingly stop/start using the interface, is this planned?

**Comment by krish2718** [comment]
Should we also extend `wifi_nm_iface_type` with `FAP` i.e., `FULL AP`?

* `SoftAP` -> WPA supplicant
* `FullAP` -> hostapd

File: modules/hostap/src/supp_events.c [hunk 6]
```diff
@@ -291,21 +311,27 @@ int supplicant_send_wifi_mgmt_ap_sta_event(void *ctx,
 					   void *data)
 {
 	struct sta_info *sta = data;
-	struct wpa_supplicant *wpa_s = ctx;
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+	struct hostapd_iface *iface = ctx;
+	char *ifname = iface->bss[0]->conf->iface;
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply modules/hostap/src/supp_events.c:321]: How about use name `"ap_ctx"` instead of `"iface"` here?

File: modules/hostap/src/supp_api.c (hunk 5 above)
Comment (nxf58150) [reply modules/hostap/src/supp_api.c:980]: Do you mean reuse the code of configuring network? I think the commands between softAP and fullAP are different.

Comment (krish2718) [reply modules/hostap/src/supp_api.c:980]: Yes.

> I think the commands between softAP and fullAP are different.

Ok, let me check the differences in commands then, thanks.

**Comment by nxf58150** [comment]
> I don't see the dynamic interfaces support for HOSTAPD_AP i.e., monitoring interface state events and accordingly stop/start using the interface, is this planned?

Currently, we are not using dynamic interface support on hostapd. We need time to do research on this. How about we keep it like this for now? Will do code enhancement in the future.

**Comment by krish2718** [comment]
> > I don't see the dynamic interfaces support for HOSTAPD_AP i.e., monitoring interface state events and accordingly stop/start using the interface, is this planned?
> 
> Currently, we are not using dynamic interface support on hostapd. We need time to do research on this. How about we keep it like this for now? Will do code enhancement in the future.

But other drivers might still use it (Nordic driver uses it, but still not upstream). May be add a Kconfig dependency on `depends on !WIFI_NM_WPA_SUPPLICANT_INF_MON`?

**Comment by krish2718** [comment]
> > Should we also extend `wifi_nm_iface_type` with `FAP` i.e., `FULL AP`?
> > 
> > * `SoftAP` -> WPA supplicant
> > * `FullAP` -> hostapd
> 
> Setting up AP with hostapd or wpa_supplicant is transparent for user so I think extending `wifi_nm_iface_type` maybe not a necessary task. By checking the type, user will know the interface is AP, no matter it is SoftAP or FullAP. If we don't extend it, will it affect setting up AP with wpa_supplicant?

Agreed, in that case let's rename SAP ->AP and hide Soft/Full.

File: modules/hostap/src/supp_events.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (nxf58150) [reply modules/hostap/src/supp_events.c]: Renamed to `ap_ctx` here.

File: modules/hostap/src/supp_main.c (hunk 1 above)
Comment (nxf58150) [reply modules/hostap/src/supp_main.c]: Removed line 845.

File: modules/hostap/src/supp_main.c (hunk 2 above)
Comment (nxf58150) [reply modules/hostap/src/supp_main.c]: Updated.

**Comment by nxf58150** [comment]
> Isn't this depend on [zephyrproject-rtos/hostap#15](https://github.com/zephyrproject-rtos/hostap/pull/15), the west.yml should be updated too in that case.

Updated west.yml.

File: modules/hostap/Kconfig (hunk 3 above)
Comment (nxf58150) [reply modules/hostap/Kconfig]: Updated.

**Comment by nxf58150** [comment]
> Should we also extend `wifi_nm_iface_type` with `FAP` i.e., `FULL AP`?
> 
> * `SoftAP` -> WPA supplicant
> * `FullAP` -> hostapd

As discussed, will upload another PR to rename iface type from SAP to AP.

**Comment by nxf58150** [comment]
> HOSTAPD_AP

The station interface of our chip is still using WIFI_NM_WPA_SUPPLICANT_INF_MON so adding dependency of !WIFI_NM_WPA_SUPPLICANT_INF_MON for HOSTAPD_AP will have side effect.  The hostapd is not using dynamic interface support currently but it doesn't affect the functionality of WIFI_NM_WPA_SUPPLICANT_INF_MON.

File: modules/hostap/src/supp_api.c [hunk 7]
```diff
@@ -951,11 +1187,36 @@ int supplicant_ap_enable(const struct device *dev,
...
 	struct wpa_supplicant *wpa_s;
 	int ret = -1;
+#endif
 
 	k_mutex_lock(&wpa_supplicant_mutex, K_FOREVER);
 
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+	iface = get_hostapd_handle(dev);
+	if (!iface) {
+		ret = -1;
```
Comment (jerome-pouiller) [inline modules/hostap/src/supp_api.c:1203]: `ENOENT`?

File: modules/hostap/src/supp_api.h [hunk 8]
```diff
@@ -138,6 +138,17 @@ int supplicant_filter(const struct device *dev, struct wifi_filter_info *filter)
...
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+/**
+ * @brief Get Wi-Fi AP Status
+ *
+ * @param dev Wi-Fi device
+ * @param params AP status
+ * @return 0 for OK; -1 for ERROR
+ */
+int hapd_state(const struct device *dev, int *state);
+#endif
```
Comment (jerome-pouiller) [inline modules/hostap/src/supp_api.h:150]: I suggest to provide the same public API whatever the value of `CONFIG_WIFI_NM_HOSTAPD_AP`:
```
#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
int hapd_state(const struct device *dev, int *state);
#else
static inline int hapd_state(const struct device *dev, int *state) {
    return -EINVAL;
}
#endif
```

Thus the caller does not have to add `#ifdef` around the call to this function.

File: modules/hostap/src/supp_main.c [hunk 9]
```diff
@@ -523,6 +585,455 @@ static int register_supplicant_event_socket(struct supplicant_context *ctx)
...
+	bss->wps_state       = WPS_STATE_CONFIGURED;
+	bss->eap_server      = 1;
+#ifdef CONFIG_WPS
+	bss->ap_setup_locked = 1;
+#endif
+	conf->channel        = 1;
+	conf->acs            = conf->channel == 0;
+#ifdef CONFIG_ACS
+	conf->acs_num_scans  = 1;
+#endif /* CONFIG_ACS */
```
Comment (jerome-pouiller) [inline modules/hostap/src/supp_main.c:856]: Don't Repeat Yourself. If the condition is obvious, I suggest to not repeat it after the `#endif`.

File: modules/hostap/src/supp_main.c [hunk 10]
```diff
@@ -523,6 +585,455 @@ static int register_supplicant_event_socket(struct supplicant_context *ctx)
...
+#ifdef CONFIG_DPP
+	struct dpp_global_config dpp_conf;
+#endif /* CONFIG_DPP */
+
+	os_memset(interfaces, 0, sizeof(struct hapd_interfaces));
+	interfaces->reload_config      = hostapd_reload_config;
+	interfaces->config_read_cb     = hostapd_config_read2;
+	interfaces->for_each_interface = hostapd_for_each_interface;
+	interfaces->ctrl_iface_init    = NULL;
+	interfaces->ctrl_iface_deinit  = NULL;
```
Comment (jerome-pouiller) [inline modules/hostap/src/supp_main.c:955]: This assignment is useless since you already called `memset()` (Probably you think it make sense to assign all the members of `interfaces`. However content of `interfaces` will probably change and the way the members are initialized will look illogical).

File: subsys/net/l2/wifi/wifi_shell.c [hunk 11]
```diff
@@ -1295,9 +1295,10 @@ static int cmd_wifi_ap_enable(const struct shell *sh, size_t argc,
 			      char *argv[])
 {
 	struct net_if *iface = net_if_get_wifi_sap();
-	static struct wifi_connect_req_params cnx_params;
+	struct wifi_connect_req_params cnx_params;
```
Comment (jerome-pouiller) [inline subsys/net/l2/wifi/wifi_shell.c:1298]: This patch change the lifetime of `cnx_params`. Can you confirm you have checked address of this variable is never copied somewhere else?

**Review by jerome-pouiller — COMMENTED** [review]
Sorry the late review.

Thank you for your effort. I hope my comments are not too late.

File: subsys/net/l2/wifi/wifi_shell.c (hunk 11 above)
Comment (nxf58150) [reply subsys/net/l2/wifi/wifi_shell.c:1298]: With `"static"`, this parameter will carry configurations from last time. For example, I first enable AP of WPA-PSK mode and set psk in cnx_params. Later, I disabled this AP and start it as open mode. There will be warning log `"Passphrase provided without security configuration"`. 
I think it is meaningless to use "static" because AP configurations might be different for each time. 
This parameter only used in this API so I simply removed the keywork "static" and do memset of this parameter.

File: modules/hostap/src/supp_main.c (hunk 9 above)
Comment (nxf58150) [reply modules/hostap/src/supp_main.c]: Removed comment after `#endif`.

File: modules/hostap/src/supp_api.c (hunk 7 above)
Comment (nxf58150) [reply modules/hostap/src/supp_api.c]: Modified `-1` to `ENOENT`.

File: modules/hostap/src/supp_api.c [hunk 12]
```diff
@@ -906,10 +955,151 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
...
+{
+	struct hostapd_iface *iface;
+	int ret = 0;
+
+	k_mutex_lock(&wpa_supplicant_mutex, K_FOREVER);
+
+	iface = get_hostapd_handle(dev);
+	if (!iface) {
+		wpa_printf(MSG_ERROR, "Device %s not found", dev->name);
+		ret = -1;
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply modules/hostap/src/supp_api.c]: Modified `-1` to `ENOENT`.

File: subsys/net/l2/wifi/wifi_shell.c [hunk 13]
```diff
@@ -1295,9 +1295,10 @@ static int cmd_wifi_ap_enable(const struct shell *sh, size_t argc,
...
 	int ret;
 
+	memset(&cnx_params, 0, sizeof(struct wifi_connect_req_params));
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply subsys/net/l2/wifi/wifi_shell.c]: Updated to struct `wifi_connect_req_params cnx_params = {0};`

File: modules/hostap/src/supp_main.c (hunk 10 above)
Comment (nxf58150) [reply modules/hostap/src/supp_main.c]: Removed unnecessary assignment.

File: modules/hostap/src/supp_api.h (hunk 8 above)
Comment (nxf58150) [reply modules/hostap/src/supp_api.h:154]: Updated.

File: modules/hostap/src/supp_api.c [hunk 14]
```diff
@@ -951,11 +1187,36 @@ int supplicant_ap_enable(const struct device *dev,
 
 int supplicant_ap_disable(const struct device *dev)
 {
+#ifdef CONFIG_WIFI_NM_HOSTAPD_AP
+	struct hostapd_iface *iface;
+	int ret = 0;
+#else
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply modules/hostap/src/supp_api.c:1191]: Vendors can choose to setup AP with hostapd by enabling this flag. Or, the AP will be setup by wpa_supplicant. 
Also, I found that there are many files in Zephyr has `#ifdef` in .c files, such as net_if.c, tcp.c. So I think it might be ok if I put `#ifdef` here?

File: modules/hostap/src/supp_api.c [hunk 15]
```diff
@@ -906,10 +955,151 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
...
+			goto out;
+		}
+		iface->bss[0]->conf->wpa_key_mgmt = 0;
+	}
+
+	if (params->mfp != WIFI_MFP_DISABLE) {
+		if (!hostapd_cli_cmd_v("set ieee80211w %d", params->mfp)) {
+			goto out;
+		}
+	}
```
[the remark this answers is not in the document]
Comment (nxf58150) [reply modules/hostap/src/supp_api.c:1085]: Removed the if condition check for this parameter. Now, the `"set ieee80211w"` will always run here.

File: modules/hostap/src/supp_api.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (jerome-pouiller) [reply modules/hostap/src/supp_api.c:1191]: Obviously, conditional compilation is needed. However, I am afraid this file become a nightmare to maintain with `#ifdef` inside.

I suggest to split this file in two (or three). Then, rely on CMakeLists.txt (using `zephyr_library_sources_ifdef()`) to manage conditional compilation.

[the remark this answers is not in the document]
Comment (krish2718) [reply modules/hostap/src/supp_api.c:1191]: Sure, got it, may be `hapd_api.c` and then move all hostapd related code to that file.

File: modules/hostap/src/supp_api.c [hunk 16]
```diff
@@ -906,10 +955,151 @@ int supplicant_channel(const struct device *dev, struct wifi_channel_info *chann
...
+				goto out;
+			}
+		} else if (params->security == WIFI_SECURITY_TYPE_PSK) {
+			if (!hostapd_cli_cmd_v("set wpa 2")) {
+				goto out;
+			}
+			if (!hostapd_cli_cmd_v("set wpa_key_mgmt WPA-PSK")) {
+				goto out;
+			}
+			if (!hostapd_cli_cmd_v("set wpa_passphrase %s", params->psk)) {
```
[the remark this answers is not in the document]
Comment (krish2718) [reply modules/hostap/src/supp_api.c]: Yes, in `supp_api.c` we always use `\"%s\"` to handle the spaces.

File: modules/hostap/src/supp_api.c (hunk 14 above)
[the remark this answers is not in the document]
Comment (gannicus-gj) [reply modules/hostap/src/supp_api.c:1191]: Hi @krish2718 
if split supp_api.c to hapd_api.c, do you think supp_main.c and supp_events.c needs same split?

the reason I said a lot work to do is on our internal code base, because  following feature which already finished, such as: uap enterpise,  ap dpp are all done based on supp_api.c (none split file),
so there will be a lot of rebase work to do in our internal code base

do you think it is possible that we first finish upstream all our PRs in this none split file, and finally move to a split file
for supp_api.c, supp_events.c and supp_main.c?

[the remark this answers is not in the document]
Comment (krish2718) [reply modules/hostap/src/supp_api.c:1191]: Fine with me, we can add an improvement in GH to track.

[the remark this answers is not in the document]
Comment (gannicus-gj) [reply modules/hostap/src/supp_api.c:1191]: Hi @jerome-pouiller , do you agree with our above plan plan? could you pls take time to comment?
thanks in advance

[the remark this answers is not in the document]
Comment (gannicus-gj) [reply modules/hostap/src/supp_api.c:1191]: Hi @jerome-pouiller 
any feedback/comments on following plan ?
 if split supp_api.c to hapd_api.c, do you think supp_main.c and supp_events.c needs same split?

the reason I said a lot work to do is on our internal code base, because target to upstream features which already finished, such as: uap enterpise, ap dpp are all done based on supp_api.c (none split file), so there will be a lot of rebase work to do in our internal code base

do you think it is possible that we first finish upstream all our PRs in this none split file, and finally move to a split file for supp_api.c, supp_events.c and supp_main.c in one Pull Request ?

**Comment by jukkar** [comment]
There is quite weird error in CI
```
CONFLICT (content): Merge conflict in west.yml
error: could not apply 971c7cdaba8... manifest: Update hostap to latest
hint: Resolve all conflicts manually, mark them as resolved with
hint: "git add/rm <conflicted_files>", then run "git rebase --continue".
hint: You can instead skip this commit: run "git rebase --skip".
hint: To abort and get back to the state before "git rebase", run "git rebase --abort".
Could not apply 971c7cdaba8... manifest: Update hostap to latest
```
Could you rebase this PR against latest upstream, and force push, perhaps it solves this issue.

File: modules/hostap/src/supp_api.h [hunk 17]
```diff
@@ -190,6 +190,21 @@ int supplicant_get_wifi_conn_params(const struct device *dev,
...
+/**
+ * @brief Get Wi-Fi AP Status
+ *
+ * @param dev Wi-Fi device
+ * @param params AP status
+ * @return 0 for OK; -1 for ERROR
+ */
+int hapd_state(const struct device *dev, int *state);
+#else
+int hapd_state(const struct device *dev, int *state) {
```
Comment (jukkar) [inline modules/hostap/src/supp_api.h:203]: This function should be static inline as suggested by @jerome-pouiller

Comment (nxf58150) [reply modules/hostap/src/supp_api.h]: Updated. Thanks!

**Review by jukkar — DISMISSED** [review]
LGTM

**Comment by jukkar** [comment]
@jerome-pouiller & @krish2718 please take a look

File: modules/hostap/CMakeLists.txt [hunk 18]
```diff
@@ -209,11 +210,31 @@ zephyr_library_compile_definitions_ifdef(CONFIG_WIFI_NM_WPA_SUPPLICANT_AP
 	CONFIG_NO_VLAN
 	CONFIG_NO_ACCOUNTING
 	NEED_AP_MLME
+	CONFIG_IEEE80211AC
```
Comment (krish2718) [inline modules/hostap/CMakeLists.txt:213]: why is this needed? As this is independent to how the AP is being setup, it should be in a separate commit.

Comment (nxf58150) [reply modules/hostap/CMakeLists.txt:213]: Our chip supports 11AC(both STA and AP) and is enabled by default. Without this flag, 11AC related source code won't be compiled and set conf->ieee80211ac is useless.

Comment (krish2718) [reply modules/hostap/CMakeLists.txt:213]: ok, can you please put this in a separate commit.

Comment (nxf58150) [reply modules/hostap/CMakeLists.txt:213]: Updated.

**Review by krish2718 — DISMISSED** [review]

**Review by jerome-pouiller — DISMISSED** [review]

**Comment by jukkar** [comment]
@nxf58150 the hostap PR is now merged, please update manifest file and re-submit this one

**Comment by nxf58150** [comment]
> @nxf58150 the hostap PR is now merged, please update manifest file and re-submit this one

Updated manifest file.

**Review by jukkar — APPROVED** [review]

**Review by jerome-pouiller — APPROVED** [review]

**Review by krish2718 — APPROVED** [review]

**Review by krish2718 — APPROVED** [review]
